src/core/config/mle.h

```diff
@@ -277,7 +277,8 @@
  *
  */
 #ifndef OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE
-#define OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE 0
+#define OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE \
+    (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2 && OPENTHREAD_FTD)
```

Comment:
Note that we shouldn't define a default option by `OPENTHREAD_FTD`, `OPENTHREAD_MTD` or `OPENTHREAD_RADIO`. The reason is that they (FTD, MTD, RADIO) are only visible to OpenThread core. When building other components, like the platform, `OPENTHREAD_FTD` is not defined and thus evaluates to 0. And it emits warning. See the logs from the failed CI build:
```
/home/runner/work/openthread/openthread/examples/../examples/platforms/simulation/radio.c: In function ‘generateAckIeData’:
/home/runner/work/openthread/openthread/examples/../src/core/config/mle.h:281:67: warning: "OPENTHREAD_FTD" is not defined [-Wundef]
     (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2 && OPENTHREAD_FTD)
                                                                   ^
/home/runner/work/openthread/openthread/examples/../examples/platforms/simulation/radio.c:1156:5: note: in expansion of macro ‘OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE’
 #if OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE
     ^
/home/runner/work/openthread/openthread/examples/../examples/platforms/simulation/radio.c: At top level:
/home/runner/work/openthread/openthread/examples/../src/core/config/mle.h:281:67: warning: "OPENTHREAD_FTD" is not defined [-Wundef]
     (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2 && OPENTHREAD_FTD)
                                                                   ^
/home/runner/work/openthread/openthread/examples/../examples/platforms/simulation/radio.c:1242:5: note: in expansion of macro ‘OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE’
 #if OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE
     ^
  CC       libopenthread_simulation_a-trel.o
  CC       libopenthread_simulation_a-uart.o
  CC       virtual_time/libopenthread_simulation_a-platform-sim.o
  CC       virtual_time/libopenthread_simulation_a-alarm-sim.o
  AR       libopenthread-simulation.a
ar: `u' modifier ignored since `D' is the default (see `U')
make[5]: Nothing to be done for 'all-am'.
Making all in apps
Making all in cli
  CXX      ot_cli_ftd-cli_uart.o
  CXX      ot_cli_mtd-cli_uart.o
  CC       ot_cli_mtd-main.o
  CC       ot_cli_ftd-main.o
  CXXLD    ot-cli-ftd
  CXXLD    ot-cli-mtd
../../../src/core/libopenthread-ftd.a(libopenthread_ftd_a-link_metrics.o): In function `ot::LinkMetrics::LinkMetrics::ConfigureEnhAckProbing(ot::LinkMetrics::EnhAckFlags, ot::LinkMetrics::Metrics const&, ot::Neighbor&) [clone .part.15]':
/home/runner/work/openthread/openthread/examples/../src/core/radio/radio.hpp:649: undefined reference to `otPlatRadioConfigureEnhAckProbing'
```
`otPlatRadioConfigureEnhAckProbing` is defined in platform and wrapped by `OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE`. When building it, `OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE` evaluates to `0` because there is no definition of `OPENTHREAD_FTD`.

We actually met this issue before. And we elimiated the warnings by not defining default options y thins like `OPENTHREAD_FTD`. See #5923 and #5428

In this case, we can remove `OPENTHREAD_FTD` here and replace the occurences of guard `OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE` to `OPENTHREAD_FTD && OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE`. We must replace that in `ot-nrf528xx` first. I apologize I didn't do this before.

Reply:
@Irving-cl Thanks for the info. 

The problem with using `OPENTHREAD_FTD && OPENTHREAD_CONFIG_MLE_LINK_METRICS_SUBJECT_ENABLE` is that it will disallowing Link Metrics Subject feature on MTD? Can MTD has Link Metrics Subject?

@librasungirl @Irving-cl

Reply:
Checked Spec and found Spec defines that `Link Metrics Probing Subject` is MUST for Router and REED, but not defined for FED, MED and SED.